### ISSUE TITLE
Rule 950140 improve

### ISSUE THREAD

**Issue by leveryd** [coreruleset/coreruleset#3266]
```
SecRule RESPONSE_BODY "@rx ^#\!\s?/" \
    "id:950140,\
    phase:4,\
    block,\
    capture,\
    t:none,\
    msg:'CGI source code leakage',\
    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-disclosure',\
    tag:'paranoia-level/1',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/118/116',\
    tag:'PCI/6.5.6',\
    ctl:auditLogParts=+E,\
    ver:'OWASP_CRS/4.0.0-rc1',\
    severity:'ERROR',\
    setvar:'tx.outbound_anomaly_score_pl1=+%{tx.error_anomaly_score}'"
```

The above rule make false positive alarm because ^ does not only match the first line of the response body, i think it should be fixed to below

```
SecRule RESPONSE_BODY "@rx ^(.{10})" \
    "id:950140,\
    phase:4,\
    block,\
    capture,\
    t:none,\
    msg:'CGI source code leakage',\
    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    tag:'application-multi',\
    tag:'language-multi',\
    tag:'platform-multi',\
    tag:'attack-disclosure',\
    tag:'paranoia-level/1',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/118/116',\
    tag:'PCI/6.5.6',\
    ctl:auditLogParts=+E,\
    ver:'OWASP_CRS/3.3.2',\
    severity:'ERROR',\
    setvar:'tx.first_ten_chars=%{tx.1}',\
    chain"
    SecRule TX:FIRST_TEN_CHARS "@rx ^#\!\s?/" \
      "setvar:'tx.outbound_anomaly_score_pl1=+%{tx.error_anomaly_score}',\
      setvar:'tx.anomaly_score_pl1=+%{tx.error_anomaly_score}'"
```

You can refer https://github.com/SpiderLabs/ModSecurity/issues/2921

**Comment by M4tteoP** [comment]
Hi @leveryd, thank you for your report. `^` is indeed matching the start of a new line, because the engine comes with `multiline` enabled, therefore the match might not necessarily be at the first line of the response.

You found this rule as a potential FPs source, or are you currently facing false positives? Are they coming from a very specific use case that might require writing an exception for it? Would it be possible to have an example of a legit response that matches this rule?

Adding to the loop @airween after reading his knowledgable answers in https://github.com/SpiderLabs/ModSecurity/issues/2921

As a side note, `RESPONSE_BODY "@rx ^` is a pattern currently used also across several `RESPONSE-955-WEB-SHELLS` rules, even if they are less generic patterns compared to `950140`.

**Comment by leveryd** [comment]
Yes, it makes false positive alarm in my situation.

And i think the rule was going to match the first line, but it does not actually.

**Comment by M4tteoP** [comment]
I dug a bit more about it.

> `^` is indeed matching the start of a new line, because the engine comes with `multiline` enabled, therefore the match might not necessarily be at the first line of the response.

This sentence I previously wrote is not completely true, it rather depends on the engine or on how ModSec has been compiled.

**ModSec2 (v2.9.7) (for Apache):**
I’m not an expert of this codebase, but to the best of what I understood, `PCRE_MULTILINE` flag is:

- Always enabled for some operators (like [gsbLookup](https://github.com/SpiderLabs/ModSecurity/blob/v2.9.7/apache2/re_operators.c#L1691-L1696) and [verifyCC](https://github.com/SpiderLabs/ModSecurity/blob/v2.9.7/apache2/re_operators.c#L2810-L2816)): `PCRE2_DOTALL | PCRE2_MULTILINE`

-  Not present on the `rx` operator that rather has: `PCRE2_DOTALL | PCRE2_DOLLAR_ENDONLY`
-  Under `standalone/regex.c` there also is a logic that includes `PCRE_MULTILINE` if `cflags` comes with `AP_REG_NEWLINE`. I guess it is at compilation time. But on this file, there is not even mention of PCRE2, which should have been added to 2.9.* 🧐

**Libmodsecurity (v.3.0.9) (E.g. Nginx):**
- Comes with `pcre2_options = (PCRE2_DOTALL|PCRE2_MULTILINE)`. See [here](https://github.com/SpiderLabs/ModSecurity/blob/v3.0.9/src/utils/regex.cc#L68).

**Coraza (v3.0.2) (tested on coraza-caddy):**
- Since [#732](https://github.com/corazawaf/coraza/pull/732), rx operator has been aligned to Libmodsecurity with both [dotall and multiline](https://github.com/corazawaf/coraza/blob/0dabae31d831f4f6d0f1bb57a1c5adc5cd4f2eb5/internal/operators/rx.go#L29):  `"(?sm)%s", options.Arguments`

Tests confirm it:
Adding `SecRule RESPONSE_BODY "@rx ^}" "id:911,phase:4, deny, status:405, capture,log, t:none, msg:'multiline works'"`
With a request like: `curl 'localhost:80/anything'`
Usual response:
```
{
  "args": {},
  "data": "",
  "files": {},
  "form": {},
  "headers": {
    "Accept": "*/*",
    "Connection": "close",
    "Host": "localhost",
    "User-Agent": "curl/7.87.0"
  },
  "json": null,
  "method": "GET",
  "origin": "192.168.48.1",
  "url": "http://localhost/anything"
} 
```
Results:
- Apache does not trigger this rule
- Nginx triggers the rule, and writes the log, but can not enforce the deny (headers already sent)
- Caddy-Coraza triggers the rule and denies the request

**Recap / tl;dr:**
**multiline** is not always enabled, regexes might have it or not, depending on the engine. Should the CRS take into account this and provide rules with the same behavior regardless of `multiline`? Should we align the engines?

Could you please provide us with the waf engine and version you are using? @leveryd

**Comment by dune73** [comment]
See #3277 for full coverage of this problem across multiple engines.
